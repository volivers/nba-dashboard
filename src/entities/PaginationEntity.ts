import { Pagination } from '../interfaces/Pagination';

export interface PaginationEntity {
  current_page: number;
  next_page: number;
  per_page: number;
}

export const deserializePaginationEntity = (entity: PaginationEntity): Pagination => ({
  currentPage: entity.current_page,
  nextPage: entity.next_page,
  perPage: entity.per_page,
});

export const mockedPaginationEntity: PaginationEntity = {
  current_page: 1,
  next_page: 2,
  per_page: 25,
};
