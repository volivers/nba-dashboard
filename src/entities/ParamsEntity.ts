export interface ParamsEntity {
  page?: number;
  per_page?: number;
  game_ids?: number[];
  player_ids?: number[];
  team_ids?: number[];
}
