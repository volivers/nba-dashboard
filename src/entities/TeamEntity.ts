import { Team } from '../interfaces/Team';

export interface TeamEntity {
  abbreviation: string;
  city: string;
  conference: string;
  division: string;
  full_name: string;
  id: number;
  name: string;
}

export const deserializeTeamEntity = (entity: TeamEntity): Team => ({
  abbreviation: entity?.abbreviation,
  city: entity?.city,
  conference: entity?.conference,
  division: entity?.division,
  fullName: entity?.full_name,
  id: entity?.id,
  name: entity?.name,
});

export const mockedHomeTeamEntity: TeamEntity = {
  abbreviation: 'BOS',
  city: 'Boston',
  conference: 'Conference',
  division: 'Division',
  full_name: 'Boston Celtics',
  id: 2,
  name: 'Celtics',
};

export const mockedAwayTeamEntity: TeamEntity = {
  abbreviation: 'LAC',
  city: 'Los Angeles',
  conference: 'Conference',
  division: 'Division',
  full_name: 'LA Clippers',
  id: 1,
  name: 'Clippers',
};