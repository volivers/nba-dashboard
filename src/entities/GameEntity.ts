import { Game } from '../interfaces/Game';
import { TeamEntity, deserializeTeamEntity, mockedAwayTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface GameEntity {
  date: string;
  home_team: TeamEntity;
  home_team_score: number;
  id: number;
  period: number;
  postseason: boolean;
  season: number;
  status: string;
  time: string;
  visitor_team: TeamEntity;
  visitor_team_score: number;
}

export const deserializeGameEntity = (entity: GameEntity): Game => ({
  date: entity.date,
  homeTeam: deserializeTeamEntity(entity.home_team),
  homeTeamScore: entity.home_team_score,
  id: entity.id,
  period: entity.period,
  postseason: entity.postseason,
  season: entity.season,
  status: entity.status,
  time: entity.time,
  visitorTeam: deserializeTeamEntity(entity.visitor_team),
  visitorTeamScore: entity.visitor_team_score,
});

export const mockedGameEntity: GameEntity = {
  date: '2000-11-01T00:00:00.000Z',
  home_team: mockedHomeTeamEntity,
  home_team_score: 100,
  id: 1,
  period: 4,
  postseason: true,
  season: 1,
  status: 'Final',
  time: '',
  visitor_team: mockedAwayTeamEntity,
  visitor_team_score: 80,
};
