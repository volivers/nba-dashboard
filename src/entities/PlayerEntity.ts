import { Player } from '../interfaces/Player';
import { TeamEntity, deserializeTeamEntity, mockedHomeTeamEntity } from './TeamEntity';

export interface PlayerEntity {
  first_name: string;
  height_feet?: number;
  height_inches?: number;
  id: number;
  last_name: string;
  position: string;
  team: TeamEntity;
  weight_pounds?: number;
}

export const deserializePlayerEntity = (entity: PlayerEntity): Player => ({
  firstName: entity.first_name,
  heightFeet: entity?.height_feet,
  heightInches: entity?.height_inches,
  id: entity.id,
  lastName: entity.last_name,
  position: entity.position,
  team: deserializeTeamEntity(entity.team),
  weightPounds: entity?.weight_pounds,
});

export const mockedPlayerEntity: PlayerEntity = {
  first_name: 'Neemias',
  id: 88,
  last_name: 'Queta',
  position: 'C',
  team: mockedHomeTeamEntity,
};
