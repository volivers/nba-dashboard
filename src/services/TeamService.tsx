import axios from 'axios';
import { TeamEntity, deserializeTeamEntity } from '../entities/TeamEntity';
import { PaginationEntity, deserializePaginationEntity } from '../entities/PaginationEntity';
import { Team } from '../interfaces/Team';
import { Pagination } from '../interfaces/Pagination';
import { Params, serializeParams } from '../interfaces/Params';

export interface TeamResponse {
  data: Team[];
  meta: Pagination;
}

export interface TeamResponseEntity {
  data: TeamEntity[];
  meta: PaginationEntity;
}

export const deserializeTeamResponse = (entity: TeamResponseEntity): TeamResponse => ({
  data: entity.data.map(deserializeTeamEntity),
  meta: deserializePaginationEntity(entity.meta)
});

export const TEAMS_API_URL = process.env.REACT_APP_API_URL + '/teams'

export class TeamService {
  public static async getAll(params?: Params): Promise<TeamResponse> {
    const response = await axios<TeamResponseEntity>({
      method: 'GET',
      url: TEAMS_API_URL,
      params: { ...serializeParams(params) },
      headers: {
        'Content-Type': 'application/json',
        'X-RapidAPI-Key': process.env.REACT_APP_API_KEY,
        'X-RapidAPI-Host': 'free-nba.p.rapidapi.com'
      }
    });

    return deserializeTeamResponse(response.data);
  };
};
