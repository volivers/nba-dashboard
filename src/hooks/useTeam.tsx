import { useQuery } from 'react-query';
import { TeamService } from '../services/TeamService';

export const useTeam = () => useQuery('teams', () => TeamService.getAll(), { keepPreviousData: true });
