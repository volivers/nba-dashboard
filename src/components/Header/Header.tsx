import React, { useMemo } from 'react';
import { Navbar, Container, Button, ButtonGroup, Offcanvas } from 'react-bootstrap';
import { useTeam } from '../../hooks/useTeam';
import Logo from '../Logo/Logo';

export interface HeaderProps {
  activeTeamId: number;
  onTeamChange: (id: number) => void;
}

const Header = ({ activeTeamId, onTeamChange }: HeaderProps) => {
  const { data: results, isLoading } = useTeam();

  const activeTeam = useMemo(() => results?.data.find(({ id }) => id === activeTeamId), [activeTeamId, results?.data]);

  return (
    <Navbar expand={false} className="bg-body-tertiary mb-3 p-3" bg="dark" data-bs-theme="dark" sticky="top">
      <Container fluid>
        <Navbar.Brand><img src="img/NBA.svg" alt="nba-logo" loading="lazy" /></Navbar.Brand>
        {!isLoading && (
          <Navbar.Text as="h1" className="p-0">
            <Logo name={activeTeam?.abbreviation} />
            {activeTeam?.fullName}
          </Navbar.Text>
        )}
        <Navbar.Toggle className="border-0" />
        <Navbar.Offcanvas aria-label="offcanvasNavbarLabel" bg="dark" data-bs-theme="dark" placement="end">
          <Offcanvas.Header closeButton>
            <Offcanvas.Title as="h1">Teams</Offcanvas.Title>
          </Offcanvas.Header>
          <Offcanvas.Body>
            <ButtonGroup vertical>
              {results?.data?.map(({ abbreviation, fullName, id }) => (
                <Button
                  className="text-light text-start text-decoration-none"
                  disabled={activeTeamId === id}
                  onClick={() => onTeamChange(id)}
                  key={id}
                  variant="link"
                >
                  <Logo name={abbreviation} />
                  {fullName}
                </Button>
              ))}
            </ButtonGroup>
          </Offcanvas.Body>
        </Navbar.Offcanvas>
      </Container>
    </Navbar>
  );
};

export default Header;