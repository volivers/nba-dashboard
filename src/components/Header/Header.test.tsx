import { render } from '@testing-library/react';
import Header, { HeaderProps } from './Header';
import { deserializePaginationEntity, mockedPaginationEntity } from '../../entities/PaginationEntity';
import { TeamResponse } from '../../services/TeamService';
import { useTeam } from '../../hooks/useTeam';
import { deserializeTeamEntity, mockedHomeTeamEntity } from '../../entities/TeamEntity';

jest.mock('../../hooks/useTeam');
const mockedUseTeam = useTeam as jest.Mock;

describe('Header', () => {
  let props: HeaderProps;
  let mockedTeamResponse: TeamResponse;

  beforeEach(() => {
    props = {
      activeTeamId: 2,
      onTeamChange: jest.fn(),
    };

    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(),
        removeListener: jest.fn(),
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });

    mockedTeamResponse = {
      data: [deserializeTeamEntity(mockedHomeTeamEntity)],
      meta: deserializePaginationEntity(mockedPaginationEntity),
    };

    mockedUseTeam.mockResolvedValue({
      ...mockedTeamResponse,
      isLoading: false,
    });
  });

  it('renders properly', () => {
    const { container } = render(<Header {...props} />);

    expect(container).toMatchSnapshot();
  });
});
