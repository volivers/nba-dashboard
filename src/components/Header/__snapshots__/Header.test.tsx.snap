// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders properly 1`] = `
<div>
  <nav
    class="bg-body-tertiary mb-3 p-3 navbar navbar-light bg-dark sticky-top"
    data-bs-theme="dark"
  >
    <div
      class="container-fluid"
    >
      <span
        class="navbar-brand"
      >
        <img
          alt="nba-logo"
          loading="lazy"
          src="img/NBA.svg"
        />
      </span>
      <h1
        class="p-0 navbar-text"
      >
        <img
          alt="undefined-logo"
          class="mx-2"
          height="40"
          loading="lazy"
          src="img/undefined.svg"
        />
      </h1>
      <button
        aria-label="Toggle navigation"
        class="border-0 navbar-toggler collapsed"
        type="button"
      >
        <span
          class="navbar-toggler-icon"
        />
      </button>
      <div
        aria-label="offcanvasNavbarLabel"
        bg="dark"
        class="offcanvas offcanvas-end"
        data-bs-theme="dark"
      >
        <div
          class="offcanvas-header"
        >
          <h1
            class="offcanvas-title"
          >
            Teams
          </h1>
          <button
            aria-label="Close"
            class="btn-close"
            type="button"
          />
        </div>
        <div
          class="offcanvas-body"
        >
          <div
            class="btn-group-vertical"
            role="group"
          />
        </div>
      </div>
    </div>
  </nav>
</div>
`;
